### ISSUE TITLE
cargo asm ignores flags set in .cargo/config

### ISSUE THREAD

**Issue by yodaldevoid** [gnzlbg/cargo-asm#22]
I am building an embedded project that has a few flags set in a ".cargo/config" file. `cargo asm` is unable to build my project as it ignores the `rustflags` field set for the target.

**Comment by gnzlbg** [comment]
Thanks for filling the issue. 

`cargo asm` is just a script that calls `cargo`, so if `cargo` picks these flags in `.cargo/config`, `cargo asm` should too, unless one of the options passed to `cargo` is incompatible with some of the flags that are set there. 

Do you have a reproducer? Or could you upload to github a tiny project that I can use to reproduce this?

**Comment by yodaldevoid** [comment]
Looking at your code it seems like the reason that the flags set in `.cargo/config` are being ignored is because you are explicitly setting the `RUSTFLAGS` environment variable. Setting `RUSTFLAGS` overrides the settings in `.cargo/config` as per the [docs](https://doc.rust-lang.org/cargo/reference/config.html#environment-variables).

As for how to fix this, I'm not entirely sure. [`BuildContext::rustflag_args`](https://docs.rs/cargo/0.28.0/cargo/core/compiler/struct.BuildContext.html#method.rustflags_args) in the `cargo` crate might do what you need, but the documentation is lacking. If that fails, I think you could do something similar to how `cargo` does it. The [env_args](https://github.com/rust-lang/cargo/blob/13c258ecaefccaad88a173e8b8dc8517db9e8aad/src/cargo/core/compiler/build_context/mod.rs#L337-L464) function builds the flags up and the relevant functions seem to be available publicly in the `cargo` crate.

I will see if I can't get you a project to test with later today.

**Comment by gnzlbg** [comment]
> As for how to fix this, I'm not entirely sure. BuildContext::rustflag_args in the cargo crate might do what you need, but the documentation is lacking.

Just note that we are not using `cargo` as a library here. 

> Looking at your code it seems like the reason that the flags set in .cargo/config are being ignored is because you are explicitly setting the RUSTFLAGS environment variable. 

IIRC there is an alternative way of passing `RUSTFLAGS` to `cargo` using `cargo rustc ... -- ...RUSTFLAGS go here...`, maybe we could migrate to using that instead, but that `cargo` API did not have as many features as the `cargo build` API back when this was built. Hopefully that has changed. What I don't know is if `cargo rustc` respects the flags in `.cargo/config` or not.

**Comment by yodaldevoid** [comment]
> Just note that we are not using `cargo` as a library here.

> that `cargo` API did not have as many features as the `cargo build` API back when this was built.

Yes, I noticed that you were not calling `cargo` as a library and instead was calling it as a normal program. Do you know what features were missing in the past from the library API that you required?

I think I see another project forming to add to my ever growing list...

**Comment by cole-miller** [comment]
`cargo rustc` does indeed look at `.cargo/config`; for example, if you set `rustflags = ["-C", "target-cpu=native"]` (or whatever) under the `[build]` section and run `cargo rustc --verbose`, the printed invocation contains those flags (after all the ones passed by Cargo).

**Comment by lassemoldrup** [comment]
Would be nice if this was fixed. I just spent a lot of time trying to find out why the portable intrinsic I was using didn't compile to the correct instruction, only to discover it was because cargo-asm didn't use the assigned target feature.

**Comment by jimzer** [comment]
+1 to fix this. 

I also get confused a lot with the ASM ouput (when using `--target-cpu=native` in my config for instance). I don't get the proper instruction.
The only workaround I found is to repeat the flags in the environment variable just before calling cargo asm like: `RUSTFLAGS='-C target-cpu=native' cargo asm`.

Generally, I think it's a good practice to avoid configuration duplication.

In any case thanks for this crate, super useful when developping performance driven applications.
